vk/router.go

```diff
@@ -139,6 +149,52 @@ func (rt *Router) handlerWrap(inner HandlerFunc) httprouter.Handle {
 	}
 }
 
+// wsHandlerWrap returns an httprouter.Handle that uses the `inner` vk.WebSocketHandleFunc to handle the request
+//
+// inner accepts a Gorilla `Conn` and reads and writes messages to it
+//
+func (rt *Router) wsHandlerWrap(inner WebSocketHandlerFunc) httprouter.Handle {
+	return func(w http.ResponseWriter, r *http.Request, params httprouter.Params) {
+		var status int
+		var body []byte
+
+		// create a context handleWrap the configured logger
+		// (and use the ctx.Log for all remaining logging
+		// in case a scope was set on it)
+		ctx := NewCtx(rt.log, params, w.Header())
+		ctx.UseScope(defaultScope{ctx.RequestID()})
+
+		logDone := rt.logRequest(r, ctx)
+
+		upgrader := websocket.Upgrader{
+			ReadBufferSize:  1024,
+			WriteBufferSize: 1024,
+			CheckOrigin:     func(r *http.Request) bool { return true },
+		}
+
+		conn, err := upgrader.Upgrade(w, r, nil)
+		if err != nil {
+			status, body, _ = errorOrOtherToBytes(ctx.Log, err)
+			w.WriteHeader(status)
+			w.Write(body)
+
+			logDone(status)
+			return
+		}
+
+		err = inner(r, ctx, conn)
+
+		if err != nil {
+			// tbd what we'll do with the error
+			conn.Close()
+			logDone(500)
```

Comment:
Also feels a little weird to log status codes as they're essentially meaningless.

Reply:
You should be sending a webscoket status code not HTTP

https://www.rfc-editor.org/rfc/rfc6455.html#section-7.4


Also use CloseError as opposed to Close 

https://pkg.go.dev/github.com/gorilla/websocket#CloseError

Reply:
What I've just learned with some research is that Gorilla doesn't actually implement graceful shutdowns for you, and you have to more or less roll it yourself, which is a little unfortunate. `CloseError` is just a type rather than a function that handles it for you. This package does implement it nicely (can just call `Close` with the code and a message and it'll handle it all for you) https://pkg.go.dev/nhooyr.io/websocket

But it's not nearly as popular as Gorilla. Do we go through the effort of implementing the `Close` song and dance ourselves?

Reply:
That's an interesting observation, I'm a bit surprised that Gorilla doesn't expose a means to execute a server initiated close handshake with a specific reason code. 

The referenced package seems popular enough but it hasn't had a ton of activity recently which is a bit worrisome. 

I'd prefer not to roll our own personally but I'll leave that for you to decide.

Reply:
I may be misinterpreting this but further investigation seems to suggest the close handshake is client initiated and the server just closes the TCP connection if it wants to abort the session. 

Close Control Frame
https://www.rfc-editor.org/rfc/rfc6455.html#section-5.5.1

Server Initiated Close
https://www.rfc-editor.org/rfc/rfc6455.html#section-7.2.2

https://www.rfc-editor.org/rfc/rfc6455.html#section-7.1.1

Reply:
In conclusion, I was wrong carry on as you were. Resolving comment